File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/dynamic_programming/house_robber/P198_HouseRobber.py
from typing import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        return self.dp(nums)

    def recursive(self, nums, i):
        if i == len(nums):
            return 0
        # rob
        p1 = nums[i] + self.recursive(nums, i + 2)
        # not rob
        p2 = self.recursive(nums, i + 1)

    def dp(self, nums):
        n = len(nums) + 2
        dp = [0 for i in range(n)]
        for i in range(n - 3, -1, -1):
            dp[i] = max(nums[i] + dp[i + 2], dp[i + 1])
        return dp[0]

    def dp2(self, nums):
        n = len(nums) + 2
        p, p1, p2 = 0, 0, 0
        for i in range(n - 3, -1, -1):
            p = max(nums[i] + p2, p1)
            p2 = p1
            p1 = p
        return p

File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/dynamic_programming/house_robber/P198_HouseRobber.java
package com.dsa.leetcode.labuladong.dynamic_programming.house_robber;

/**
 * @since 2023/11/30 22:16
 */
public class P198_HouseRobber {

    public int rob(int[] nums) {
        if (nums == null || nums.length == 0) {
            return 0;
        }
        return dp2(nums);
    }

    private int recursive(int[] nums, int i) {
        if (i >= nums.length) {
            return 0;
        }
        // rob
        int p1 = nums[i] + recursive(nums, i + 2);
        // not rob
        int p2 = recursive(nums, i + 1);
        return Math.max(p1, p2);
    }

    private int dp(int[] nums) {
        int n = nums.length + 2;
        int[] dp = new int[n];
        // 从右到左
        for (int i = n - 3; i >= 0; i--) {
            dp[i] = Math.max(nums[i] + dp[i + 2], dp[i + 1]);
        }
        return dp[0];
    }

    private int dp2(int[] nums) {
        int p = 0;
        int p1 = 0;
        int p2 = 0;
        for (int i = nums.length - 1; i >= 0; i--) {
            p = Math.max(nums[i] + p2, p1);
            p2 = p1;
            p1 = p;
        }
        return p;
    }

    public static void main(String[] args) {
        int[] nums = {2, 7, 9, 3, 1};
        P198_HouseRobber p198 = new P198_HouseRobber();
        System.out.println(p198.rob(nums));
    }
}
